### PR TITLE
Align AD inference with training normalization

### LINKED ISSUE #40: AD diffusion: inference bypasses training-time instance normalization
## Area

`ad_diffusion`

## Environment

- OS: Linux 5.15.0-1110-azure
- Python: 3.12.13
- Revision inspected: `da2c4d0`
- Installation: source checkout with the `ad_diffusion` package on `PYTHONPATH`

## What happened

The diffusion denoiser receives different input scales during training and inference.

`TSDiffuser_base.calc_loss()` applies per-window instance normalization before constructing the noisy and conditioned denoiser inputs:

```python
data_mean = observed_data.mean(dim=(1, 2), keepdim=True)
data_std = observed_data.std(dim=(1, 2), keepdim=True) + 1e-5
observed_data = (observed_data - data_mean) / data_std
observed_data = torch.clamp(observed_data, min=-10.0, max=10.0)
```

However, both `TSDiffuser_base.evaluate()` and `TSDiffuser_base.evaluate_with_dpm()` pass the raw SDK-preprocessed `observed_data` directly to `impute()` / `dpm_solver_impute()`.

For the same deterministic window:

```python
observed_data = torch.tensor(
    [[[0.0, 5.0, 10.0, 15.0], [20.0, 25.0, 30.0, 35.0]]]
)
```

the training path presents standardized values to the denoiser, while the inference path presents the original `0..35` values. The mismatch occurs before thresholding or anomaly-score post-processing and affects both standard diffusion inference and DPM-Solver inference.

## Expected behavior

Training and inference should use the same input-scale contract at the denoiser boundary. Returned reconstructions should remain in the SDK target scale so residual computation continues comparing like-for-like values.

## Impact

The published inference paths can evaluate the denoiser on a distribution different from the one used by `calc_loss()`. This affects reconstruction values and therefore the residual anomaly scores consumed by downstream threshold strategies.

### PR DESCRIPTION
**Pull request by PansulBhatt**
Fixes #40

## Summary

- Reuse the per-window normalization already used by `calc_loss()` for both standard and DPM-Solver inference.
- Denormalize generated samples before returning them so SDK residuals continue comparing reconstructions and targets in the same external scale.
- Add regression coverage for the denoiser-boundary scale contract and returned reconstruction scale.

## Root cause

`TSDiffuser_base.calc_loss()` standardizes each observed window before constructing the conditioned/noisy denoiser input, but `evaluate()` and `evaluate_with_dpm()` previously passed raw SDK-preprocessed values directly into `impute()` and `dpm_solver_impute()`.

As a result, the same observation reached the denoiser in different numerical distributions during training and inference. The mismatch happened before residual scoring or thresholding and affected both inference implementations.

## Implementation

The change extracts the existing training transform into `_normalize_observed_data()` and uses it in all three paths. Inference retains the per-window mean/std and applies `_denormalize_samples()` before returning generated samples. This preserves the public SDK contract: `recon` and `target` remain directly comparable in the preprocessed target scale.

## Public reconstruction benchmark

To avoid threshold-strategy effects, reconstruction quality was measured directly from `inference["recon"]` against `inference["target"]`:

```python
error = recon - target
mae = np.abs(error).mean()
mse = np.square(error).mean()
rmse = np.sqrt(mse)
```

Protocol:

- Public complete TSB-AD-M series: `002_MSL_id_1_Sensor_tr_500_1st_900.csv` and `144_SMAP_id_1_Sensor_tr_2052_1st_5300.csv`.
- Published NV-Tesseract AD checkpoint and `curriculum_medium.yaml`.
- Fixed seed `2026`, `nsample=1`, DPM-Solver with 5 steps.
- No SCS/MACS thresholding, point adjustment, top-k selection, or anomaly-label-dependent threshold.
- `PR #39` denotes the complementary-mask aggregation fix; `both` applies that change together with this PR.

| Dataset | Variant | MAE | MSE | RMSE | MAE reduction vs baseline | MSE reduction vs baseline |
| --- | --- | ---: | ---: | ---: | ---: | ---: |
| MSL | baseline | 2.982 | 14.157 | 3.763 | — | — |
| MSL | this PR | 0.968 | 1.484 | 1.218 | 67.6% | 89.5% |
| MSL | PR #39 | 0.339 | 0.784 | 0.886 | 88.6% | 94.5% |
| MSL | both | 0.184 | 0.165 | 0.407 | 93.8% | 98.8% |
| SMAP | baseline | 2.869 | 12.989 | 3.604 | — | — |
| SMAP | this PR | 0.253 | 0.133 | 0.365 | 91.2% | 99.0% |
| SMAP | PR #39 | 0.405 | 1.170 | 1.082 | 85.9% | 91.0% |
| SMAP | both | 0.050 | 0.022 | 0.149 | 98.3% | 99.8% |

The improvement also holds when errors are separated by labeled timestamp type:

| Dataset | Variant | Normal MAE | Normal MSE | Anomaly MAE | Anomaly MSE |
| --- | --- | ---: | ---: | ---: | ---: |
| MSL | baseline | 2.983 | 14.161 | 2.896 | 13.493 |
| MSL | this PR | 0.968 | 1.484 | 0.918 | 1.392 |
| MSL | PR #39 | 0.340 | 0.786 | 0.306 | 0.554 |
| MSL | both | 0.184 | 0.166 | 0.174 | 0.144 |
| SMAP | baseline | 2.869 | 12.996 | 2.867 | 12.948 |
| SMAP | this PR | 0.248 | 0.130 | 0.289 | 0.153 |
| SMAP | PR #39 | 0.404 | 1.165 | 0.411 | 1.201 |
| SMAP | both | 0.049 | 0.022 | 0.057 | 0.025 |

Residual-based anomaly precision/recall/F1 are intentionally not used as the reconstruction-quality acceptance criterion here because they depend on downstream score separation and threshold calibration.

## Validation

- `make lint`
- `make spdx-check`
- `OAIPKG_DISABLE_META_MISSING=1 PYTHONPATH=$PWD/ad_diffusion python -m pytest -q sdk/tests` (`8 passed`)
- Combined regression run with this PR and PR #39 tests (`6 passed`)

### CODE REVIEW

**Comment by agautam478** [comment]
No blocking findings. 
The change correctly reuses the training-time per-window normalization for both inference paths and denormalizes returned samples so SDK residuals stay in the external input scale.

**Review by neptunel — APPROVED** [review]
lgtm
